src/lib/storage/value_segment.cpp

```diff
@@ -6,68 +6,89 @@
 namespace opossum {
 
 template <typename T>
-ValueSegment<T>::ValueSegment(bool nullable) {
-  // Implementation goes here
-}
+ValueSegment<T>::ValueSegment(bool nullable) : _is_nullable{nullable} {}
 
 template <typename T>
 AllTypeVariant ValueSegment<T>::operator[](const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  if (is_null(chunk_offset)) {
+    return NULL_VALUE;
+  }
+  return _values.at(chunk_offset);
 }
 
 template <typename T>
 bool ValueSegment<T>::is_null(const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _is_nullable && _null_values.at(chunk_offset);
 }
 
 template <typename T>
 T ValueSegment<T>::get(const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  const auto optional = get_typed_value(chunk_offset);
+  if (!optional.has_value()) {
+    throw std::logic_error{"value is NULL"};
+  }
+  return optional.value();
 }
 
 template <typename T>
 std::optional<T> ValueSegment<T>::get_typed_value(const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  const auto variant = operator[](chunk_offset);
+  if (variant_is_null(variant)) {
+    return std::nullopt;
+  }
+  const auto value = type_cast<T>(variant);
+  return std::make_optional(value);
 }
 
 template <typename T>
 void ValueSegment<T>::append(const AllTypeVariant& value) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  const auto is_null = variant_is_null(value);
+  if (is_null && is_nullable()) {
+    _null_values.push_back(true);
+    _values.push_back(T{});
+  } else {
+    if (is_null) {
+      throw std::logic_error{"cannot append null value to non-nullable ValueSegment"};
+    }
+
+    try {
+      auto typed_value = type_cast<T>(value);
+      _values.push_back(typed_value);
+      if (is_nullable()) {
+        _null_values.push_back(false);
+      }
+    } catch (const boost::wrapexcept<boost::bad_lexical_cast>& e) {
+      throw std::logic_error{e.what()};
+    }
+  }
 }
 
 template <typename T>
 ChunkOffset ValueSegment<T>::size() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _values.size();
 }
 
 template <typename T>
 const std::vector<T>& ValueSegment<T>::values() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _values;
 }
 
 template <typename T>
 bool ValueSegment<T>::is_nullable() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _is_nullable;
 }
 
 template <typename T>
 const std::vector<bool>& ValueSegment<T>::null_values() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  if (!is_nullable()) {
+    throw std::logic_error{"ValueSegment is not nullable"};
+  }
```

Comment:
Same with Assert here :)
```suggestion
  Assert(is_nullable(), "ValueSegment is not nullable");
```